src/frequenz/sdk/timeseries/_ringbuffer.py

```diff
@@ -0,0 +1,430 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Any, Generic, List, Sequence, TypeVar
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = TypeVar("Container", list, np.ndarray)
+
+
+@dataclass
+class Gap:
+    """A gap defines the range for which we haven't received values yet."""
+
+    start: datetime
+    """Start of the range, inclusive."""
+    end: datetime
+    """End of the range, exclusive."""
+
+    def contains(self, timestamp: datetime):
```

Comment:
FYI you could also implement the `__contains__()` method to allow writing: `if date in gap: do_whatever()`.

Reply:
That seems like a good idea